File: app/src/main/java/rynkbit/tk/coffeelist/admin/item/ManageItemsAdapter.java
package rynkbit.tk.coffeelist.admin.item;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.TextView;

import org.w3c.dom.Text;

import java.util.LinkedList;
import java.util.List;

import rynkbit.tk.coffeelist.R;
import rynkbit.tk.coffeelist.db.entity.Item;

/**
 * Created by michael on 11/14/16.
 */
public class ManageItemsAdapter extends RecyclerView.Adapter {
    private ManageItemsController mController;
    private List<Item> items;

    public ManageItemsAdapter(ManageItemsController controller) {
        mController = controller;
        items = new LinkedList<>();
    }

    public void setItems(List<Item> items){
        if(items != null) {
            this.items = items;
            this.notifyDataSetChanged();
        }
    }

    public class ManageItemsHolder extends RecyclerView.ViewHolder{
        View view;
        TextView itemName;
        TextView itemDetail;

        ImageButton itemMenuButton;

        public ManageItemsHolder(View itemView) {
            super(itemView);

            view = itemView;
            itemName = (TextView) itemView.findViewById(R.id.manage_list_item_title);
            itemDetail = (TextView) itemView.findViewById(R.id.manage_list_item_detail);
            itemMenuButton = (ImageButton) itemView.findViewById(R.id.btnListItem);
        }
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater
                .from(parent.getContext())
                .inflate(R.layout.admin_manage_list_item, parent, false);
        ManageItemsHolder holder = new ManageItemsHolder(view);
        return holder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        Item item = items.get(position);
        ManageItemsHolder itemsHolder = (ManageItemsHolder) holder;

        itemsHolder.itemName.setText(item.getName());
        itemsHolder.itemDetail.setText(
                String.format(itemsHolder.view.getContext().getString(
                        R.string.item_detail)
                    , item.getPrice(), item.getStock())
        );
        itemsHolder.itemMenuButton.setOnClickListener(
                new ManageItemsMenuClickListener(mController, item));
    }

    @Override
    public int getItemCount() {
        return items.size();
    }
}

File: app/src/main/java/rynkbit/tk/coffeelist/ui/admin/item/ManageItemsAdapter.kt
package rynkbit.tk.coffeelist.ui.admin.item

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import androidx.core.widget.addTextChangedListener
import androidx.recyclerview.widget.RecyclerView
import rynkbit.tk.coffeelist.R
import rynkbit.tk.coffeelist.ui.entity.UIItem

class ManageItemsAdapter(
        private val onUpdateItemName: ((UIItem) -> Unit)?,
        private val onUpdateItemPrice: ((UIItem) -> Unit)?,
        private val onUpdateItemStock: ((UIItem) -> Unit)?,
        private val onRemoveItem: ((item: UIItem) -> Unit)?
) : RecyclerView.Adapter<ManageItemsAdapter.ViewHolder>() {
    private val items = mutableListOf<UIItem>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(
                LayoutInflater.from(parent.context)
                        .inflate(R.layout.edit_item, parent, false))
    }

    override fun getItemCount(): Int = items.size

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = items[position]

        holder.apply{
            txtId.text = item.id.toString()
            editName.setText(item.name)
            editPrice.setText(item.price.toString())
            editStock.setText(item.stock.toString())

            editName.addTextChangedListener { s -> onUpdateItemName?.apply {
                this(UIItem(item.id, s.toString(), item.price, item.stock)) }
            }
            editPrice.addTextChangedListener { onUpdateItemPrice?.apply {
                val price = editPrice.text.toString().toDoubleOrNull()
                this(UIItem(item.id, item.name, price ?: item.price, item.stock)) }
            }
            editStock.addTextChangedListener { onUpdateItemStock?.apply {
                val stock = editStock.text.toString().toIntOrNull()
                this(UIItem(item.id, item.name, item.price, stock ?: item.stock)) }
            }
            btnRemove.setOnClickListener {
                onRemoveItem?.apply { this(item) }
            }
        }
    }

    fun updateItems(itemList: List<UIItem>) {
        items.clear()
        items.addAll(itemList)
        notifyDataSetChanged()
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val txtId: TextView = itemView.findViewById(R.id.txtId)
        val editName: EditText = itemView.findViewById(R.id.editName)
        val editPrice: EditText = itemView.findViewById(R.id.editPrice)
        val editStock: EditText = itemView.findViewById(R.id.editStock)
        val btnRemove: Button = itemView.findViewById(R.id.btnRemoveItem)
    }
}